File: integ/fixtures/resource/wiki_backend.ts
// The TypeScript-source spelling of wiki_backend.mjs, loaded through
// Node's type stripping: strip-only, so no parameter properties, enums or
// namespaces here. Same two classes, same behaviour as the .py and .mjs
// twins.
import { createHash } from 'node:crypto'
import {
  Accessor,
  type CommandIO,
  ContentType,
  eisdir,
  enoent,
  enotdir,
  FileStat,
  FileType,
  GenericResource,
  type PathSpec,
  streamFromBytes,
} from '@struktoai/mirage-core'

type Pages = Record<string, string>

const PAGES: Pages = { 'notes.md': 'agents just speak bash\n' }
const FEED: Pages = { 'status.md': 'All systems go.\n' }
const ENC = new TextEncoder()
const DEC = new TextDecoder()

class PageAccessor extends Accessor {
  pages: Pages
  constructor(pages: Pages) {
    super()
    this.pages = pages
  }
}

function key(path: PathSpec): string {
  return path.resourcePath.replace(/^\/+|\/+$/g, '')
}

function readdir(accessor: PageAccessor, path: PathSpec): Promise<string[]> {
  if (key(path) !== '') throw enotdir(path)
  const parent = path.virtual.replace(/\/+$/, '')
  return Promise.resolve(
    Object.keys(accessor.pages)
      .sort()
      .map((name) => `${parent}/${name}`),
  )
}

function readBytes(accessor: PageAccessor, path: PathSpec): Promise<Uint8Array> {
  const name = key(path)
  if (name === '') throw eisdir(path)
  if (!Object.hasOwn(accessor.pages, name)) throw enoent(path)
  return Promise.resolve(ENC.encode(accessor.pages[name]))
}

function stat(accessor: PageAccessor, path: PathSpec): Promise<FileStat> {
  const name = key(path)
  const trimmed = path.virtual.replace(/\/+$/, '')
  const base = trimmed.slice(trimmed.lastIndexOf('/') + 1) || '/'
  if (name === '') {
    return Promise.resolve(new FileStat({ name: base, size: null, type: FileType.DIRECTORY }))
  }
  if (!Object.hasOwn(accessor.pages, name)) throw enoent(path)
  const data = ENC.encode(accessor.pages[name])
  const fingerprint = createHash('sha256').update(data).digest('hex').slice(0, 16)
  return Promise.resolve(
    new FileStat({
      name: base,
      size: data.length,
      type: FileType.FILE,
      content: ContentType.TEXT,
      fingerprint,
    }),
  )
}

function write(accessor: PageAccessor, path: PathSpec, data: Uint8Array): Promise<void> {
  const name = key(path)
  if (name === '' || name.includes('/')) throw enotdir(path)
  accessor.pages[name] = DEC.decode(data)
  return Promise.resolve()
}

function makeIO(): CommandIO<PageAccessor> {
  return {
    readdir,
    readBytes,
    readStream: (a, p, i) => streamFromBytes(readBytes, a, p, i),
    stat,
    write,
    isMounted: () => true,
    local: false,
  }
}

// Owned content: the pages ride the state and rebuild without help. The
// registry constructs a referenced class with the mount's config object,
// so the constructor reads its pages off that shape.
export class WikiResource extends GenericResource<PageAccessor> {
  readonly store: PageAccessor

  constructor(config: { pages?: Pages } = {}) {
    const store = new PageAccessor({ ...(config.pages ?? PAGES) })
    super({ name: 'wiki', accessor: store, io: makeIO(), supportsSnapshot: true })
    this.store = store
  }

  override getState(): { type: string; pages: Pages } {
    return { type: this.kind, pages: { ...this.store.pages } }
  }

  override loadState(state: { type: string; pages?: Pages }): void {
    this.store.pages = { ...(state.pages ?? {}) }
  }
}

// Observed content: the default state asks to be handed back live.
export class FeedResource extends GenericResource<PageAccessor> {
  constructor() {
    super({ name: 'feed', accessor: new PageAccessor(FEED), io: makeIO(), supportsSnapshot: true })
  }
}

File: integ/fixtures/resource/wiki_backend.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

# A backend a deployment ships as a file and names from yaml
# (`resource: ./wiki_backend.py:WikiResource`). Two classes over one
# page store, one per half of the versioning design: WikiResource owns
# its pages and carries them in its state, so a snapshot rebuilds the
# mount through the recorded reference with the pages as they were;
# FeedResource keeps the default state, so a load has to be handed the
# live resource and refuses otherwise. The TypeScript twins beside this
# file must behave identically, because the point of the ref form is
# that one deployment runs on both hosts.

import hashlib
from copy import deepcopy
from functools import partial

from mirage import (NULL_INDEX, Accessor, CommandIO, ContentType, FileStat,
                    FileType, GenericResource, IndexCacheStore, PathSpec,
                    stream_from_bytes)

PAGES = {"notes.md": "agents just speak bash\n"}
FEED = {"status.md": "All systems go.\n"}


class PageAccessor(Accessor):

    def __init__(self, pages: dict[str, str]) -> None:
        self.pages = pages


def _key(path: PathSpec) -> str:
    return path.resource_path.strip("/")


async def readdir(accessor: PageAccessor,
                  path: PathSpec,
                  index: IndexCacheStore = NULL_INDEX) -> list[str]:
    if _key(path):
        raise NotADirectoryError(path.virtual)
    parent = path.virtual.rstrip("/")
    return [f"{parent}/{name}" for name in sorted(accessor.pages)]


async def read_bytes(accessor: PageAccessor,
                     path: PathSpec,
                     index: IndexCacheStore = NULL_INDEX) -> bytes:
    key = _key(path)
    if not key:
        raise IsADirectoryError(path.virtual)
    if key not in accessor.pages:
        raise FileNotFoundError(path.virtual)
    return accessor.pages[key].encode()


async def stat(accessor: PageAccessor,
               path: PathSpec,
               index: IndexCacheStore = NULL_INDEX) -> FileStat:
    key = _key(path)
    name = path.virtual.rstrip("/").rsplit("/", 1)[-1] or "/"
    if not key:
        return FileStat(name=name, size=None, type=FileType.DIRECTORY)
    if key not in accessor.pages:
        raise FileNotFoundError(path.virtual)
    data = accessor.pages[key].encode()
    return FileStat(name=name,
                    size=len(data),
                    type=FileType.FILE,
                    content=ContentType.TEXT,
                    fingerprint=hashlib.sha256(data).hexdigest()[:16])


async def write(accessor: PageAccessor, path: PathSpec, data: bytes) -> None:
    key = _key(path)
    if not key or "/" in key:
        raise NotADirectoryError(path.virtual)
    accessor.pages[key] = data.decode()


def make_io() -> CommandIO:
    return CommandIO(readdir=readdir,
                     read_bytes=read_bytes,
                     read_stream=partial(stream_from_bytes, read_bytes),
                     stat=stat,
                     write=write,
                     is_mounted=lambda a: True,
                     local=False)


class WikiResource(GenericResource):
    """Owned content: the pages ride the state and rebuild without help."""

    def __init__(self, pages: dict[str, str] | None = None) -> None:
        self.store = PageAccessor(deepcopy(PAGES if pages is None else pages))
        super().__init__(name="wiki",
                         accessor=self.store,
                         io=make_io(),
                         supports_snapshot=True)

    def get_state(self) -> dict:
        return {"type": self.name, "pages": deepcopy(self.store.pages)}

    def load_state(self, state: dict) -> None:
        self.store.pages = deepcopy(state.get("pages", {}))


class FeedResource(GenericResource):
    """Observed content: the default state asks to be handed back live."""

    def __init__(self) -> None:
        super().__init__(name="feed",
                         accessor=PageAccessor(FEED),
                         io=make_io(),
                         supports_snapshot=True)
